// @manuth/culture-info/lib/CultureInfo.js
/**
 * Provides information about a specific culture (called a locale for unmanaged code development).
 */
export class CultureInfo {
    /**
     * Initializes a new instance of the {@link CultureInfo `CultureInfo`} class.
     *
     * @param name
     * A non case-sensitive predefined {@link CultureInfo `CultureInfo`} name.
     */
    constructor(name) {
        /**
         * The script of the culture.
         */
        this.script = null;
        /**
         * The region of the culture.
         */
        this.region = null;
        let result = /^(?:(\w{2})(?:[-_](\w{4}))?(?:[-_](\w{2}))?)?$/g.exec(name);
        if (result === null) {
            throw new RangeError("Invalid language-tag.");
        }
        else if (result[0].length === 0) {
            this.Language = "";
        }
        else {
            this.Language = result[1].toLowerCase();
            if (result[2]) {
                this.Script = result[2].charAt(0).toUpperCase() + result[2].slice(1).toLocaleLowerCase();
            }
            if (result[3]) {
                this.Region = result[3].toUpperCase();
            }
        }
    }
    /**
     * Gets the {@link CultureInfo `CultureInfo`} object that is culture-independent (invariant).
     */
    static get InvariantCulture() {
        if (!CultureInfo.invariantCulture) {
            CultureInfo.invariantCulture = new CultureInfo("");
        }
        return CultureInfo.invariantCulture;
    }
    /**
     * Gets a value indicating whether the current {@link CultureInfo `CultureInfo`} represents a neutral culture.
     */
    get IsNeutralCulture() {
        return !this.Region;
    }
    /**
     * Gets the culture name.
     */
    get Name() {
        let result = this.Language;
        if (this.Script) {
            result += "-" + this.Script;
        }
        if (this.Region) {
            result += "-" + this.Region;
        }
        return result;
    }
    /**
     * Gets the {@link CultureInfo `CultureInfo`} that represents the parent culture of the current {@link CultureInfo `CultureInfo`}.
     */
    get Parent() {
        if (this.Script === null && this.Region === null) {
            return CultureInfo.InvariantCulture;
        }
        else {
            let nameParts = this.Name.split("-");
            return new CultureInfo(nameParts.slice(0, nameParts.length - 1).join("-"));
        }
    }
    /**
     * Gets or sets the language of the culture.
     */
    get Language() {
        return this.language;
    }
    /**
     * @inheritdoc
     */
    set Language(value) {
        this.language = value;
    }
    /**
     * Gets or sets the script of the culture.
     */
    get Script() {
        return this.script;
    }
    /**
     * @inheritdoc
     */
    set Script(value) {
        this.script = value;
    }
    /**
     * Gets or sets the region of the culture.
     */
    get Region() {
        return this.region;
    }
    /**
     * @inheritdoc
     */
    set Region(value) {
        this.region = value;
    }
    /**
     * Returns a string which represents the object.
     *
     * @returns
     * A string which represents the object.
     */
    toString() {
        return this.Name;
    }
    /**
     * Returns the primitive value of the object.
     *
     * @returns
     * The primitive value of the object.
     */
    valueOf() {
        return this.Name;
    }
}
